src/connector/src/macros.rs

```diff
@@ -13,152 +13,225 @@
+        $crate::dispatch_source_enum! {SplitImpl, { $impl }, {$inner_name, $prop_type_name, IgnoreSplitType}, $body}
+    }};
+}
 
-            fn restore_from_json(value: JsonbVal) -> Result<Self> {
-                let mut value = value.take();
-                let json_obj = value.as_object_mut().unwrap();
-                let split_type = json_obj.remove(SPLIT_TYPE_FIELD).unwrap().as_str().unwrap().to_string();
-                let inner_value = json_obj.remove(SPLIT_INFO_FIELD).unwrap();
-                Self::restore_from_json_inner(&split_type, inner_value.into())
-            }
+#[macro_export]
+macro_rules! impl_split {
+    ({$({ $variant_name:ident, $prop_name:ty, $split:ty}),*}) => {
+
+        #[derive(Debug, Clone, EnumAsInner, PartialEq, Hash)]
+        pub enum SplitImpl {
+            $(
+                $variant_name($split),
+            )*
         }
 
-        impl SplitImpl {
-            pub fn get_type(&self) -> String {
-                match self {
-                    $( Self::$variant_name(_) => $connector_name, )*
-                }
-                    .to_string()
-            }
+        $(
+            impl TryFrom<SplitImpl> for $split {
+                type Error = anyhow::Error;
 
-            pub fn update_in_place(&mut self, start_offset: String) -> anyhow::Result<()> {
-                match self {
-                    $( Self::$variant_name(inner) => inner.update_with_offset(start_offset)?, )*
+                fn try_from(split: SplitImpl) -> std::result::Result<Self, Self::Error> {
+                    match split {
+                        SplitImpl::$variant_name(inner) => Ok(inner),
+                        other => Err(anyhow::anyhow!("expect {} but get {:?}", stringify!($split), other))
+                    }
                 }
-                Ok(())
             }
 
-            pub fn encode_to_json_inner(&self) -> JsonbVal {
-                match self {
-                    $( Self::$variant_name(inner) => inner.encode_to_json(), )*
+            impl From<$split> for SplitImpl {
+                fn from(split: $split) -> SplitImpl {
+                    SplitImpl::$variant_name(split)
                 }
             }
 
-            fn restore_from_json_inner(split_type: &str, value: JsonbVal) -> Result<Self> {
-                match split_type.to_lowercase().as_str() {
-                    $( $connector_name => <$split>::restore_from_json(value).map(SplitImpl::$variant_name), )*
-                        other => {
-                    Err(anyhow!("connector '{}' is not supported", other))
-                    }
-                }
-            }
-        }
+        )*
     }
 }
 
+#[macro_export]
+macro_rules! dispatch_source_prop {
+    ($impl:expr, $source_prop:tt, $body:expr) => {{
+        use $crate::source::ConnectorProperties;
+        $crate::dispatch_source_enum! {ConnectorProperties, { $impl }, {$source_prop, IgnorePropType, IgnoreSplitType}, {$body}}
```

Comment:
> Is there a way to show the expanded code of a macro_rules! without calling it just like a function definition

You can write an example in a random function and use `cargo expand` to see the expanded code.

> coud we flatten these macros a bit

This main idea of this PR is to declare all source related information in only one place so that when developing a new source, we don't have to declare these information in places previously scattered around in our codebase.